File: src/schemas/event.ts
// FILE: src/schemas/event.ts
// CaptureEvent — the universal event format.
// Cross-language contract: mirrors Go struct in daemon/internal/capture/event.go.
// The Go daemon writes CaptureEvent JSON to ~/.unfade/events/YYYY-MM-DD.jsonl;
// the TypeScript CLI reads and validates them with this schema.

import { z } from "zod";

export const CaptureEventSchema = z.object({
  id: z.string().uuid(),
  projectId: z.string(),
  timestamp: z.string().datetime(),
  source: z.enum(["git", "ai-session", "terminal", "browser", "manual", "mcp-active"]),
  type: z.enum([
    "commit",
    "diff",
    "branch-switch",
    "revert",
    "stash",
    "merge-conflict",
    "ai-conversation",
    "ai-completion",
    "ai-rejection",
    "command",
    "error",
    "retry",
    "bookmark",
    "tab-visit",
    "annotation",
  ]),
  content: z.object({
    summary: z.string(),
    detail: z.string().optional(),
    files: z.array(z.string()).optional(),
    branch: z.string().optional(),
    project: z.string().optional(),
  }),
  gitContext: z
    .object({
      repo: z.string(),
      branch: z.string(),
      commitHash: z.string().optional(),
    })
    .optional(),
  metadata: z.record(z.string(), z.unknown()).optional(),
});

export type CaptureEvent = z.infer<typeof CaptureEventSchema>;

File: daemon/internal/capture/event.go
// FILE: daemon/internal/capture/event.go
// CaptureEvent — Go struct mirroring src/schemas/event.ts (Zod).
// Any changes here MUST be synchronized with the TypeScript schema.
// The daemon writes these as JSON to ~/.unfade/events/YYYY-MM-DD.jsonl.

package capture

// CaptureEvent is the universal event format written by the daemon.
// Field names use camelCase JSON tags to match the TypeScript schema exactly.
type CaptureEvent struct {
	ID         string         `json:"id"`
	ProjectID  string         `json:"projectId"`
	Timestamp  string         `json:"timestamp"`
	Source     string         `json:"source"` // git | ai-session | terminal | browser | manual | mcp-active
	Type       string         `json:"type"`   // commit | diff | branch-switch | ...
	Content    EventContent   `json:"content"`
	GitContext *GitContext    `json:"gitContext,omitempty"`
	Metadata   map[string]any `json:"metadata,omitempty"`
}

// EventContent holds the primary payload of a capture event.
type EventContent struct {
	Summary string   `json:"summary"`
	Detail  string   `json:"detail,omitempty"`
	Files   []string `json:"files,omitempty"`
	Branch  string   `json:"branch,omitempty"`
	Project string   `json:"project,omitempty"`
}

// GitContext provides repository context for git-sourced events.
type GitContext struct {
	Repo       string `json:"repo"`
	Branch     string `json:"branch"`
	CommitHash string `json:"commitHash,omitempty"`
}
